Add options to document list component

andysellick opened the pull request:
Adds the following...

Add 'subtext' option, small grey text to appear beneath metadata. This is needed for search results in finders, (where we want to use the document list component). This is not styled other than to match the font size of the description, specific styles will be added in finder-frontend for any elements that will be passed into the subtext area.

![Screen Shot 2019-06-11 at 10 29 39](https://user-images.githubusercontent.com/861310/59260607-da961380-8c33-11e9-9ef6-9b7f9258eae7.png)

Add 'highlight' option, including optional highlight text. This is needed to highlight the 'top result' in finders.

![Screen Shot 2019-06-11 at 16 20 16](https://user-images.githubusercontent.com/861310/59284824-d33b2e00-8c64-11e9-9a39-e1c40d911d86.png)

Trello card: https://trello.com/c/DAblviIN/793-investigate-using-more-components-in-finder-frontend

Some examples of what this looks like currently:

- [subtext](https://www.gov.uk/search/all?parent=&keywords=%22every+child+matters%22&level_one_taxon=&manual=&public_timestamp%5Bfrom%5D=&public_timestamp%5Bto%5D=&order=relevance) (was broken but fixed in [this PR](https://github.com/alphagov/finder-frontend/pull/1198))
- [highlight option](https://www.gov.uk/find-eu-exit-guidance-business?parent=&keywords=eori&order=relevance) can be seen here, but you need to set a request header of `GOVUK-ABTest-FinderAnswerABTest` to `B` (I'd suggest using [this chrome extension](https://chrome.google.com/webstore/detail/modheader/idgpnmonknjnojddfkpgkljpfnnfcklj)). In the real world it wouldn't have the metadata beneath the description text, I just left this in for the component guide to make sure the layout still worked okay.

vanitabarrett commented:
@andysellick Have you got links to finders with these on? Specifically, the subtext one

vanitabarrett commented on app/assets/stylesheets/govuk_publishing_components/components/_document-list.scss:
```diff
@@ -43,6 +43,12 @@
 
+.gem-c-document-list__subtext {
+  @include govuk-font(14);
+  margin: govuk-spacing(1) 0 0 0;
```
Could this just be `margin-top: govuk-spacing(1)` instead?

andysellick replied:
No - the paragraph seems to be inheriting bottom margin from somewhere, which I wanted to remove.

vanitabarrett commented on app/views/govuk_publishing_components/components/_document_list.html.erb:
```diff
@@ -14,7 +14,10 @@
   <% end %>
     <% items.each do |item| %>
-      <li class="gem-c-document-list__item<%= within_multitype_list_class %> <%= brand_helper.brand_class %>">
+      <li class="gem-c-document-list__item<%= within_multitype_list_class %> <%= brand_helper.brand_class %> <%= 'gem-c-document-list__item--highlight' if item[:highlight] %>">
```
Should we follow the pattern we use for `within_multitype_list_class` and construct the class at the top of the file to avoid too much logic within the HTML itself?

andysellick replied:
Yes, have changed now, thanks.

andysellick commented:
> @andysellick Have you got links to finders with these on? Specifically, the subtext one

I haven't, unfortunately. The only way I know what these things look like is by putting debug code into my local so it outputs. But I'll try to find some examples when I look at addressing the reviews.

andysellick commented:
> I'm wondering if the subtext addition could be structured as part of `gem-c-document-list__item-metadata` (potentially with a modifier for the grey font colour) instead of a new element (`gem-c-document-list__subtext`)

@alex-ju that's an approach I hadn't considered probably because of the way the data in `finder-frontend` gets assembled before it (would be) passed to the component. I'm in no huge rush to merge this, shall we chat about it offline? Would appreciate your input.

andysellick commented:
> I'm wondering if the subtext addition could be structured as part of `gem-c-document-list__item-metadata` (potentially with a modifier for the grey font colour) instead of a new element (`gem-c-document-list__subtext`)

@alex-ju just to summarise the conversation we just had - I'm going to leave the subtext option in place, but remove the styling from the component, as the appearance of this is specific to the application I'm going to use it in (finder-frontend) and there's additional styling already in finder-frontend for the debug stuff that will also be inside the subtext element.

andysellick commented:
@vanitabarrett turns out the subtext 'First published by' is broken 😱 but it's being fixed in [this PR](https://github.com/alphagov/finder-frontend/pull/1198) 🎉

andysellick commented:
@vanitabarrett the 'First published by' is fixed and deployed, I've added an example link to the PR but it's also live at the link below. I've removed the styling for the subtext option as it feels more appropriate for it to go in finder-frontend (there's going to be more than just this inside this element).

Are you happy with this PR?

https://www.gov.uk/search/all?parent=&keywords=%22every+child+matters%22&level_one_taxon=&manual=&public_timestamp%5Bfrom%5D=&public_timestamp%5Bto%5D=&order=relevance

vanitabarrett commented:
@andysellick Looks good to me 👍 The only comment I have is maybe the first commit message needs to be changed to reflect the fact that you're not styling it to match the existing look, and that styles will be applied within finder-frontend?
